### PR TITLE
Vehicle installation requirements (part 2) [RDY]

### LINKED ISSUE #16407: Need info on what level of lifting equipment needed for vehicle modification
Currently, "lifting equipment" is in red with no indication that you need a better / stronger jack

### PR DESCRIPTION
**Pull request by mugling**
Replaces #16941 and follows #17208

demo
- Display `LIFTING` or `JACKING` requirements (fixes #16407)
- Display installation time
- Allow specifying installation and removal requirements in JSON

### CODE REVIEW

File: src/veh_type.cpp [hunk 1]
```diff
@@ -347,8 +386,60 @@ void vpart_info::finalize()
...
+                part.removal_reqs.qualities = { { { quality_id( "WRENCH" ), 1, 1 } } };
+            } else if( part.has_flag( "TOOL_SCREWDRIVER" ) ) {
+                part.install_reqs.qualities = { { { { quality_id( "SCREW" ), 1, 1 } } } };
+                part.removal_reqs.qualities = { { { { quality_id( "SCREW" ), 1, 1 } } } };
+            } else if( part.has_flag( "NAILABLE" ) ) {
+                part.install_reqs.qualities = { { { { quality_id( "HAMMER" ), 1, 1 } } } };
+                part.removal_reqs.qualities = { { { { quality_id( "HAMMER" ), 1, 1 } } } };
+                part.install_reqs.components.push_back( { { { "nail", 20 } } } );
+            } else {
+                part.install_reqs.components.push_back( { { { "duct_tape", 20 } } } );
```
Comment (Coolthulhu) [inline src/veh_type.cpp]: Will it allow welding?

File: src/veh_type.h [hunk 2]
```diff
@@ -127,6 +132,33 @@ class vpart_info
...
+        bool legacy = true;
+
+        /** Installation requirements for this component */
+        requirement_data install_reqs;
+
+        /** Required skills to install this component */
+        std::map<skill_id, int> install_skills;
+
+        /** Installation time (in moves) for component (@see install_time), default 1 hour */
+        int install_moves = 60000;
```
Comment (Coolthulhu) [inline src/veh_type.h]: > default 1 hour

Use `HOURS( 1 )` macro then. Makes it more readable.

File: src/veh_type.cpp [hunk 3]
```diff
@@ -347,8 +386,60 @@ void vpart_info::finalize()
...
+
+        for( auto &e : part.removal_skills ) {
+            if( !e.first.is_valid() ) {
+                debugmsg( "vehicle part %s has unknown removal skill %s", part.id.c_str(), e.first.c_str() );
+            }
+        }
+
+        part.install_reqs.check_consistency( part.id.c_str() );
+
+        part.removal_reqs.check_consistency( part.id.c_str() );
```
Comment (Coolthulhu) [inline src/veh_type.cpp]: Why `c_str()` instead of just the regular `str()`?

File: src/veh_type.cpp (hunk 1 above)
Comment (mugling) [inline src/veh_type.cpp]: Yes, but this is specified via the vehicle parts requirement data

File: src/veh_type.h (hunk 2 above)
Comment (mugling) [inline src/veh_type.h]: Causes the value to be misculated.
I suppose `HOURS( 1 ) * 100` could work but that's no more readable given the comment above

Comment (Coolthulhu) [inline src/veh_type.h]: `HOURS( 100 )`

It's better to write readable code than to write unreadable code and comment it.

Comment (codemime) [inline src/veh_type.h]: The task (convert time to moves) isn't rare in the code. I suppose it deserves yet another macro (`constexpr` function) `MOVES( HOURS( 1 ) )`.

Comment (mugling) [inline src/veh_type.h]: That implies 100 hours which is definitely misleading

Comment (mugling) [inline src/veh_type.h]: @codemime could work

File: src/veh_type.cpp (hunk 1 above)
Comment (mugling) [inline src/veh_type.cpp]: See 13ddd1c. Heavy duty frames now require much more welding than standard frames

**Comment by mugling** [comment]
There's a lot more refactoring possible within `veh_interact.cpp` however that can wait for a further PR.

See updated preamble for screenshot

**Comment by Coolthulhu** [comment]
Could it also fix the regression with wheels not showing their sizes in install menu?
They did show everything before the var size tampering.

**Comment by mugling** [comment]
Preferably #17195 and #17061 first as we can then have much more interesting installation requirements. For example headlights should take a small amount of `electronics` skill as well as a supply of electrical wire.

**Comment by Coolthulhu** [comment]
But that's a rather significant regression brought by recent changes to the vehicle system.
In general, fixing regressions gets a priority over new features.

**Comment by mugling** [comment]
@Coolthulhu We were both typing at the same time - that isn't the reply to your earlier comment

**Comment by mugling** [comment]
See 360d00a. Can handle `set of 4" casters` as well as `17" wheel`

File: data/json/vehicleparts/frames.json [hunk 4]
```diff
@@ -84,7 +97,20 @@
...
     "color": "dark_gray",
     "durability": 1600,
-    "difficulty": 1,
+    "difficulty": 3,
+    "requirements": {
+      "install": {
+        "skills": [ [ "mechanics", 3 ] ],
+        "time": 240000,
+        "qualities": [ { "id": "GLARE", "level": 2 } ],
+        "tools": [ [ [ "welder", 200 ], [ "welder_crude", 300 ], [ "toolset", 150 ], [ "oxy_torch", 60 ] ] ],
```
Comment (Cyrano7) [inline data/json/vehicleparts/frames.json]: Shouldn't "toolset" be 300?

**Comment by Coolthulhu** [comment]
Conflict with #17208

File: src/game.cpp [hunk 5]
```diff
@@ -7467,16 +7464,17 @@ void game::exam_vehicle(vehicle &veh, const tripoint &p, int cx, int cy)
         // Stored in activity.index and used in the complete_vehicle() callback to finish task.
         switch (vehint.sel_cmd) {
         case 'i':
-            time = setuptime + std::max(mintime, 5000 * diff - skill * 2500);
+            time = vehint.sel_vpart_info->install_time( g->u );
```
Comment (codemime) [inline src/game.cpp]: `g->` can (should) be omitted. It's in the `game` class (x4).

File: src/game_constants.h [hunk 6]
```diff
@@ -82,7 +82,13 @@
 /** Weight per level of LIFT/JACK tool quality */
 #define TOOL_LIFT_FACTOR 500000 // 500kg/level
 
+/** Cap JACK requirements to support arbritrarily large vehicles */
+#define JACK_LIMIT 6000000 // 6000kg (6 metric tonnes)
```
Comment (codemime) [inline src/game_constants.h]: It's consistent with the rest of the definitions here, but shouldn't they become `static const`s at some point?

File: src/itype.cpp [hunk 7]
```diff
@@ -4,14 +4,25 @@
...
+#include "output.h"
 
 #include <fstream>
 #include <stdexcept>
 #include <algorithm>
 
 std::string itype::nname( unsigned int const quantity ) const
 {
-    return ngettext( name.c_str(), name_plural.c_str(), quantity );
+    std::string str( ngettext( name.c_str(), name_plural.c_str(), quantity ) );
```
Comment (codemime) [inline src/itype.cpp]: Can be `const`.

File: src/itype.cpp [hunk 8]
```diff
@@ -4,14 +4,25 @@
...
+
+    if( wheel ) {
+        return string_format( str, wheel->diameter );
+
+    } else if( engine && engine->displacement > 0 ) {
+        return string_format( str, engine->displacement / 100.0 );
```
Comment (codemime) [inline src/itype.cpp]: Rather dangerous. `string_format()` expects exact number of arguments. Why do you need to format a string using json values from the very same source? The whole idea looks like an unnecessary coupling.

File: src/veh_interact.cpp [hunk 9]
```diff
@@ -587,89 +504,73 @@ bool veh_interact::can_install_part(int msg_width){
...
+                              skill_mechanics.obj().name().c_str(), dif_steering ) << "\n";
+    }
+
+    int lvl, str;
+    quality_id qual;
+    bool use_aid, use_str;
+    item base( sel_vpart_info->item );
+    if( base.is_wheel() ) {
+        qual = JACK;
+        lvl = ceil( double( std::min( veh->total_mass() * 1000, JACK_LIMIT ) / TOOL_LIFT_FACTOR ) );
```
Comment (codemime) [inline src/veh_interact.cpp]: Missing namespace (x2).

File: src/veh_interact.cpp [hunk 10]
```diff
@@ -1140,83 +1034,71 @@ void veh_interact::do_refill()
...
-
-        std::string msg;
-
-        if( is_hand_remove ) {
-            msg = _( "You need ");
-
-        } else if( is_wood ) {
-            msg = string_format( _("You need a <color_%1$s>hammer</color> and"), status_color( has_hammer ) );
+bool veh_interact::can_remove_part( int idx ) {
+    if( sel_vpart_info == NULL ) {
```
Comment (codemime) [inline src/veh_interact.cpp]: Should be `nullptr`.

File: src/veh_interact.cpp [hunk 11]
```diff
@@ -1140,83 +1034,71 @@ void veh_interact::do_refill()
...
+        lvl = ceil( double( std::min( veh->total_mass() * 1000, JACK_LIMIT ) / TOOL_LIFT_FACTOR ) );
+        str = veh->lift_strength();
+        use_aid = max_jack >= lvl;
+        use_str = g->u.can_lift( *veh );
+    } else {
+        qual = LIFT;
+        lvl = ceil( double( base.weight() ) / TOOL_LIFT_FACTOR );
+        str = base.lift_strength();
+        use_aid = max_lift >= lvl;
+        use_str = g->u.can_lift( base );
```
Comment (codemime) [inline src/veh_interact.cpp]: I've already seen this very chunk of code above. It's rather big to duplicate it verbatim.

File: src/veh_interact.cpp [hunk 12]
```diff
@@ -2282,67 +2099,49 @@ void complete_vehicle ()
...
-            g->u.consume_tools(tools);
-        }
-        // Only parts that use charges
-        else if (is_screwable){
-            if(!has_screwdriver){
-                tools.push_back(tool_comp("duct_tape", DUCT_TAPE_USED));
-                g->u.consume_tools(tools);
-            }
+
+        case 'i': {
```
Comment (codemime) [inline src/veh_interact.cpp]: Indentation. Why the curly bracket?

File: src/veh_type.cpp [hunk 13]
```diff
@@ -412,6 +505,18 @@ std::string vpart_info::name() const
...
+int vpart_info::install_time( const Character &ch ) const {
+    ///\EFFECT_MECHANICS reduces time consumed installing vehicle parts
+    int lvl = std::min( ch.get_skill_level( skill_mechanics ).level(), MAX_SKILL );
+    return install_moves * ( 1.0 - ( lvl / 2.0 ) / MAX_SKILL );
+}
+
+int vpart_info::removal_time( const Character &ch ) const {
+    ///\EFFECT_MECHANICS reduces time consumed removing vehicle parts
+    int lvl = std::min( ch.get_skill_level( skill_mechanics ).level(), MAX_SKILL );
+    return removal_moves * ( 1.0 - ( lvl / 2.0 ) / MAX_SKILL );
```
Comment (codemime) [inline src/veh_type.cpp]: Duplicated function. I would be nice if `install_moves`/`removal_moves` could be passed as arguments to an extracted function.

File: src/game_constants.h (hunk 6 above)
Comment (mugling) [inline src/game_constants.h]: Yes on both (consistency and should be `static const`). Can be done as part of a further refactor

File: src/itype.cpp (hunk 8 above)
Comment (mugling) [inline src/itype.cpp]: `17" wheel` becoming `18" wheel` requires an update to the i18n strings whereas `%i wheel` does not. The former could get out of hand very quickly as more wheels are added.

We already have format strings in `JSON` for related reasons. It's not that you don't have a point just that the problem gets pushed elsewhere by any different solution and increasing the difficulty of i18n costs us both developers and players.

File: src/veh_interact.cpp (hunk 9 above)
Comment (mugling) [inline src/veh_interact.cpp]: Lost on that one, can you elaborate?

File: src/veh_interact.cpp (hunk 12 above)
Comment (mugling) [inline src/veh_interact.cpp]: > Indentation.

Agreed but massively increases the line count which I want to keep as low as possible for now. Can be addressed in a subsequent PR.



> Why the curly bracket?

Variable definitions within the block require it.

File: src/veh_type.cpp (hunk 13 above)
Comment (mugling) [inline src/veh_type.cpp]: Currently stub functions. No guarantee their behavior will remain equivalent. Adding a third function increases complexity for little simplification (three lines per function of which one is a doxygen comment).

File: src/veh_interact.cpp (hunk 9 above)
Comment (codemime) [inline src/veh_interact.cpp]: I mean `std::ceil()`.

File: src/itype.cpp (hunk 8 above)
Comment (BevapDin) [inline src/itype.cpp]: A single innocent '%' in the name invokes undefined behavior (format specifiers not matching the arguments). The feature is also not documented at all.

You should use a separate format string, like it's done for the morale description ("Enjoyed %i"), and replace that manually - there is a `replace_substring` function in output.h.

> It's not that you don't have a point just that the problem gets pushed elsewhere by any different solution and increasing the difficulty of i18n costs us both developers and players.

So does undefined behavior.

File: src/veh_interact.cpp (hunk 9 above)
Comment (BevapDin) [inline src/veh_interact.cpp]: The casting to double seems strange. What type has the expression `std::min( veh->total_mass() * 1000, JACK_LIMIT ) / TOOL_LIFT_FACTOR` that is casted to double? If it's already double or float, why cast it? If it's not double or float, the call to `ceil` won't do a thing. Maybe the cast should only include `TOOL_LIFT_FACTOR` or only the other side of the division?

File: src/veh_type.h [hunk 14]
```diff
@@ -127,6 +133,33 @@ class vpart_info
...
+        requirement_data removal_reqs;
+
+        /** Required skills to remove this component */
+        std::map<skill_id, int> removal_skills;
+
+        /** Removal time (in moves) for component (@see removal_time), default is half @ref install_moves */
+        int removal_moves = -1;
+
+        /** Removal time (in moves) for this component accounting for player skills */
+        int removal_time( const Character &ch ) const;
```
Comment (BevapDin) [inline src/veh_type.h]: There is pretty much code duplicated here. You have skills, moves and item requirements for installation, each duplicated for the removal.

They should be in separate class (can be a nested class) that contains those 3 things, along with the time functions and only put an install and a removal instance into the `vpart_info` itself.

You can use those class in veh_interact to reduce the duplicated printing of the requirement message.

Comment (mugling) [inline src/veh_type.h]: I also want to add repair requirements so I'm going to undertake the above consolidation then once the design is finalised.

File: src/veh_interact.cpp (hunk 11 above)
Comment (mugling) [inline src/veh_interact.cpp]: Agreed but I want to refactor jacking/lifting requirements but that's beyond the scope of this PR and can be refactored separately when those are updated

**Comment by mugling** [comment]
All changes

File: data/json/vehicleparts/frames.json [hunk 15]
```diff
@@ -41,6 +41,19 @@
...
+      "install": {
+        "skills": [ [ "mechanics", 1 ] ],
+        "time": 120000,
+        "qualities": [ { "id": "GLARE", "level": 2 } ],
+        "tools": [ [ [ "welder", 50 ], [ "welder_crude", 75 ], [ "toolset", 75 ], [ "oxy_torch", 20 ] ] ],
+        "components": [ [ [ "frame", 1 ] ] ]
+      },
+      "removal": {
+        "skills": [ [ "mechanics", 1 ] ],
+        "qualities": [ { "id": "SAW_M", "level": 1 } ]
```
Comment (Coolthulhu) [inline data/json/vehicleparts/frames.json]: Two balancing changes here:
* Wrench is no longer required. Kinda makes sense because the frame is welded on.
* You can use a multitool or a swiss army knife to remove a frame. This one sounds like a major, unintentional omission.

File: src/veh_interact.cpp [hunk 16]
```diff
@@ -2265,28 +2224,27 @@ void complete_vehicle ()
...
-        }*/ //causes runtime errors. not a critical feature; implement with PLY quality.
+
+        for( const auto& e : reqs.get_components() ) {
+            g->u.consume_items( e );
+        }
+        for( const auto& e : reqs.get_tools() ) {
+            g->u.consume_tools( e );
+        }
+
+        g->u.invalidate_crafting_inventory();
```
Comment (Coolthulhu) [inline src/veh_interact.cpp]: Unrelated, just noting:
`consume_tools` and `consume_items` should call `invalidate_crafting_inventory`.

File: src/veh_interact.cpp [hunk 17]
```diff
@@ -2297,8 +2255,8 @@ void complete_vehicle ()
         if (!broken) {
             g->m.add_item_or_charges( g->u.pos(), veh->parts[vehicle_part].properties_to_item() );
             // simple tasks won't train mechanics
-            if(type != SEL_JACK && !is_hand_remove) {
-                g->u.practice( skill_mechanics, (is_wood || is_wrenchable || is_screwable) ? 15 : 30);
+            if( reqs.get_tools().empty() ) {
+                g->u.practice( skill_mechanics, vpinfo.difficulty * 50 );
```
Comment (Coolthulhu) [inline src/veh_interact.cpp]: Looks like inverted logic here:
Before, it would train mechanics when it consumed tools.
Now it trains it when it doesn't require tools.

Also a moderate balance change.
Should probably be tied to install/remove time rather than difficulty. There is no risk of messing up, so installing and removing a complex part over and over shouldn't train the skill more than the same grind with a simple part.

File: src/veh_type.cpp [hunk 18]
```diff
@@ -171,6 +174,42 @@ void vpart_info::load( JsonObject &jo )
...
+        while( sk.has_more() ) {
+            auto cur = sk.next_array();
+            def.removal_skills.emplace( skill_id( cur.get_string( 0 ) ) , cur.size() >= 2 ? cur.get_int( 1 ) : 1 );
+        }
+
+        assign( rem, "time", def.removal_moves );
+
+        def.removal_reqs.load( rem );
+        def.legacy = false;
+    }
```
Comment (Coolthulhu) [inline src/veh_type.cpp]: `"install"` and `"removal"` handling is nearly identical, it only differs in which fields it assigns. Consider extracting the sets of fields (`removal_skills`, `removal_reqs`, `removal_moves` and the same for `install_`) into a structure and using a single function to load both of them in the exact same way.

File: data/json/vehicleparts/frames.json (hunk 15 above)
Comment (mugling) [inline data/json/vehicleparts/frames.json]: Yes, that should be `SAW_M 2` with appropriate changes in tool qualities

File: src/veh_type.cpp (hunk 18 above)
Comment (mugling) [inline src/veh_type.cpp]: Same answer as to @BevapDin - once repair requirements are also pushed to JSON we can consolidate the common parts.

There are multiple ways of doing this - one way would be to push the skills data within `requirements` but that also touches crafting code so I want to go with this in a separate PR

**Comment by Coolthulhu** [comment]
The requirements menu is very narrow, here it wraps at most requirements.

**Comment by Coolthulhu** [comment]
Batteries started costing duct tape

**Comment by mugling** [comment]
> The requirements menu is very narrow, here it wraps at most requirements.

How to make the entire dialog not waste space (it should fill the screen or a least have a much smaller border)

**Comment by mugling** [comment]
> Batteries started costing duct tape

Is this necessarily a problem and if so is it a problem right now? Overall balance needs much attention now we can have specific requirements in subsequent PR's.

**Comment by Coolthulhu** [comment]
> How to make the entire dialog not waste space (it should fill the screen or a least have a much smaller border)

Making it fullscreen would work. It's not like anyone looks at vehicle menu and other stuff at the same time.

**Comment by mugling** [comment]
> Making it fullscreen would work. It's not like anyone looks at vehicle menu and other stuff at the same time.

But how?

**Comment by Coolthulhu** [comment]
> Is this necessarily a problem and if so is it a problem right now?

Yes. It is a major regression since duct tape is rare early on. It would be better if they were free.

> But how?

I assume the window has some parameters that allow changing its x/y offsets and sizes.

Non-wrapping interface is important, but not mandatory.

> Overall balance needs much attention now we can have specific requirements in subsequent PR's.

Small regressions are tolerable.

A giant, absolutely unacceptable regression: welding is not possible, all weldable parts require duct tape.

It would be acceptable if they didn't allow duct tape and instead required welding (duct tape vehicles were always rather weird), but duct tape only vehicles are too big of a regression.

**Comment by mugling** [comment]
Ok lets adjust the legacy migration to prefer welding and have batteries require nothing other than a wrench?

**Comment by Coolthulhu** [comment]
> Ok lets adjust the legacy migration to prefer welding and have batteries require nothing other than a wrench?

Things that don't require tools in master shouldn't require them here. You aren't handling the "TOOL_NONE" flag at all.

**Comment by mugling** [comment]
Latest push includes full screen display and better battery installation requirements

**Comment by mugling** [comment]
See latest fixes

File: src/veh_type.cpp [hunk 19]
```diff
@@ -326,8 +365,65 @@ void vpart_info::finalize()
...
+                part.removal_reqs.qualities = { { { { quality_id( "SCREW" ), 1, 1 } } } };
+            } else if( part.has_flag( "NAILABLE" ) ) {
+                part.install_reqs.qualities = { { { { quality_id( "HAMMER" ), 1, 1 } } } };
+                part.removal_reqs.qualities = { { { { quality_id( "HAMMER" ), 1, 1 } } } };
+                part.install_reqs.components.push_back( { { { "nail", 20 } } } );
+            } else if( part.has_flag( "TOOL_NONE" ) ) {
+                // intentional no-op as we require nothing
+            } else {
+                part.install_reqs.qualities = { { { { quality_id( "GLARE" ), 1, 2 } } } };
+                part.install_reqs.tools.push_back( { { { "welder", 50 }, { "welder_crude", 75 }, { "oxy_torch", 25 } } } );
```
Comment (Coolthulhu) [inline src/veh_type.cpp]: Oxytorch charges should be 10.

Install requirements should also include wrench.
Removal requirements should be fine metal sawing and wrench.

File: src/veh_type.cpp [hunk 20]
```diff
@@ -326,8 +365,65 @@ void vpart_info::finalize()
...
-        auto &part = *part_ptr;
+    for( auto &vp : vehicle_part_types ) {
+        auto &part = vp.second;
+
+        // handle legacy parts without requirement data
+        if( part.legacy ) {
+            part.install_reqs.components = { { { { part.item, 1 } } } };
+
+            part.install_skills.emplace( skill_mechanics, part.difficulty );
+            part.removal_skills.emplace( skill_mechanics, part.difficulty );
```
Comment (Coolthulhu) [inline src/veh_type.cpp]: It's 2 difficulty for removal in master.
Should be 2 here, unless you want to argue balancing in this PR.

Comment (mugling) [inline src/veh_type.cpp]: Ambivalent, any preference?

Comment (Coolthulhu) [inline src/veh_type.cpp]: 2 is fine for now. Later on it could be something like `(install difficulty - 2)` or so.

Comment (mugling) [inline src/veh_type.cpp]: Or better `std::max( part.difficulty - 2, 2)`

**Comment by Coolthulhu** [comment]
Got a segfault by trying to install a part I don't have components for. Seems to be related to extra engine requirements and extra steering axles requirement, which somehow allow attempting installing a part you don't have.

**Comment by Coolthulhu** [comment]
Wait, it's not extra engines/axles, it happens every time you have no skill for the component.

**Comment by mugling** [comment]
> Wait, it's not extra engines/axles, it happens every time you have no skill for the component.

Can you rephrase?

**Comment by mugling** [comment]
Have candidate fix (flaw is in `veh_interact::can_currently_install`)

**Comment by Coolthulhu** [comment]
* Spawn casters
* Set skills to zero
* Attempt to install casters
* It works

Worse:
* Get rid of any nearby casters
* Get rid of tools necessary to install casters
* Set skills to zero
* Attempt to install casters
* debugmsgs happen, can result in segfaults

File: src/veh_interact.cpp [hunk 21]
```diff
@@ -2120,67 +2077,49 @@ void complete_vehicle ()
...
-                tools.push_back(tool_comp("duct_tape", DUCT_TAPE_USED));
-                g->u.consume_tools(tools);
-            }
+
+        case 'i': {
+        auto inv = g->u.crafting_inventory();
+
+        const auto& reqs = vpinfo.install_reqs;
+        if( !reqs.can_make_with_inventory( inv ) ) {
+           add_msg( m_info, _( "You lack the requirements to install the %s." ), vpinfo.name().c_str() );
```
Comment (Coolthulhu) [inline src/veh_interact.cpp]: "You don't meet the requirements" would sound better.

Comment (Coolthulhu) [inline src/veh_interact.cpp]: Note: no failure here, the function goes on despite lack of skills.

File: src/veh_interact.cpp [hunk 22]
```diff
@@ -2265,28 +2204,27 @@ void complete_vehicle ()
...
-                tools.push_back(tool_comp("circsaw_off", 20));
-                tools.push_back(tool_comp("oxy_torch", 10));
-                g->u.consume_tools(tools);
-            }
+    case 'o': {
+        auto inv = g->u.crafting_inventory();
+
+        const auto& reqs = vpinfo.removal_reqs;
+        if( !reqs.can_make_with_inventory( inv ) ) {
+           add_msg( m_info, _( "You lack the requirements to remove the %s." ), vpinfo.name().c_str() );
```
Comment (Coolthulhu) [inline src/veh_interact.cpp]: Same here: no failure on unmet requirements.

File: src/veh_type.cpp (hunk 20 above)
Comment (Coolthulhu) [inline src/veh_type.cpp]: Security system has flag "DIFFICULTY_REMOVE". Should be handled here too.

**Comment by mugling** [comment]
> Spawn casters
> Set skills to zero
> Attempt to install casters
> It works

1. Spawn frame
2. Spawn casters
3. Set skills to zero
4. Attempt install
5. **Doesn't work**

Did your test case have any more steps?

File: data/json/vehicleparts/frames.json [hunk 23]
```diff
@@ -84,7 +97,20 @@
...
     "color": "dark_gray",
     "durability": 1600,
-    "difficulty": 1,
+    "difficulty": 3,
+    "requirements": {
+      "install": {
+        "skills": [ [ "mechanics", 3 ] ],
+        "time": 240000,
+        "qualities": [ { "id": "GLARE", "level": 2 } ],
+        "tools": [ [ [ "welder", 200 ], [ "welder_crude", 300 ], [ "toolset", 300 ], [ "oxy_torch", 60 ] ] ],
```
Comment (Cyrano7) [inline data/json/vehicleparts/frames.json]: @mugling @Coolthulhu How should the oxy_tourch charges scale with the welder charges? I always see different ratios.

Comment (Coolthulhu) [inline data/json/vehicleparts/frames.json]: Pretty sure 1 oxy to 5 welder charges is the most common one.

Comment (mugling) [inline data/json/vehicleparts/frames.json]: To be honest I don't want to delay this PR on that ground alone

**Comment by Coolthulhu** [comment]
Lack of relevant tools isn't properly checked until the work is finished.

File: src/veh_interact.cpp (hunk 21 above)
Comment (mugling) [inline src/veh_interact.cpp]: Fixed

File: src/veh_interact.cpp (hunk 22 above)
Comment (mugling) [inline src/veh_interact.cpp]: Fixed

**Comment by mugling** [comment]
> Lack of relevant tools isn't properly checked until the work is finished.

Can you expand upon this and also confirm the above test case?

**Comment by Coolthulhu** [comment]
The extra steering axles requirement isn't properly checked when you don't meet the base requirements. For example, you can install extra casters with 0 skill if the vehicle has a steering axle.

**Comment by Coolthulhu** [comment]
> Can you expand upon this and also confirm the above test case?

You can start working on grayed out parts.

Not sure what you did differently with your test case, but in general it works like this:
* Try to install a part you don't meet some requirements for
* Some parts won't be installable
* Some parts will be installable despite unmet requirements

**Comment by mugling** [comment]
What is the **exact** sequence as I still can't reproduce that

**Comment by mugling** [comment]
The `ok -= conditional` idiom looks to be unsound

**Comment by Coolthulhu** [comment]
Minimal example would be:
* Start new game
* Drop everything
* Set skill levels to zero
* Spawn a car
* Try to install a roof over the engine
* Long action starts despite unmet dependencies

The only requirements that seem to stop it are the extra requirements like strength.

**Comment by Coolthulhu** [comment]
> The ok -= conditional idiom looks to be unsound

In a language that has implicit casts to and from `bool`, anything that treats any non-`bool` as `bool` or vice versa is generally a bad idea.
There are good reasons for why Java and C# and many other languages didn't inherit the implicit `bool` cast from C and C++.

**Comment by mugling** [comment]
Agreed, fixed in 88c5409

File: src/veh_type.cpp [hunk 24]
```diff
@@ -326,8 +365,68 @@ void vpart_info::finalize()
...
+                part.removal_reqs.qualities = { { { { quality_id( "HAMMER" ), 1, 1 } } } };
+                part.install_reqs.components.push_back( { { { "nail", 20 } } } );
+            } else if( part.has_flag( "TOOL_NONE" ) ) {
+                // intentional no-op as we require nothing
+            } else {
+                part.install_reqs.qualities = { { { { quality_id( "WRENCH" ), 1, 2 } },
+                                                  { { quality_id( "GLARE" ), 1, 2 } } } };
+                part.install_reqs.tools.push_back( { { { "welder", 50 }, { "welder_crude", 75 }, { "oxy_torch", 10 } } } );
+                part.removal_reqs.qualities = { { { { quality_id( "WRENCH" ), 1, 2 } },
+                                                  { { quality_id( "SAW_M" ), 1, 2 } } } };
```
Comment (Coolthulhu) [inline src/veh_type.cpp]: I was pretty sure wrench 2 existed, but looks like it doesn't.

Comment (mugling) [inline src/veh_type.cpp]: Ok, it should though and I'm going to handle that now

Comment (mugling) [inline src/veh_type.cpp]: Fixed

**Comment by Coolthulhu** [comment]
> Add WRENCH 2 quality

Also the survivor utility belt.

Removal difficulty still needs changing (or arguments that it is an improvement).
In master it is:
* 2 by default
* 1 for stuff that isn't welded
* Equal to install difficulty for security (which has the difficult removal flag)

**Comment by mugling** [comment]
Fixed removal difficulty

**Comment by Coolthulhu** [comment]
Screw driving 2 doesn't exist. It's fine screw driving.

**Comment by mugling** [comment]
> Screw driving 2 doesn't exist. It's fine screw driving.

Done

**Comment by Coolthulhu** [comment]
It's FINE, not F

**Comment by mugling** [comment]
Ok, sorted

File: src/veh_interact.cpp [hunk 25]
```diff
@@ -1349,20 +1318,13 @@ int veh_interact::part_at (int dx, int dy)
...
 {
-    if (g->u.has_trait("DEBUG_HS")) {
-        return true;
-    }
-    bool has_comps = crafting_inv.has_components(vpart.item, 1);
-    ///\EFFECT_MECHANICS determines which vehicle parts can be installed
-    bool has_skill = g->u.get_skill_level( skill_mechanics ) >= vpart.difficulty;
-    bool is_wheel = vpart.has_flag("WHEEL");
-    return (has_comps && (has_skill || is_wheel));
+    return g->u.has_trait( "DEBUG_HS" ) || vpart.install_reqs.can_make_with_inventory( crafting_inv );
```
Comment (BevapDin) [inline src/veh_interact.cpp]: @mugling The check for the debug trait should probably be in `can_make_with_inventory` or even deeper in the inventory. At least you have to add this special case to the check in `complete_vehicle`. Currently I can start installing things with the help of that trait, but when the character finishes it gets the sarcastic "You lack the requirements to install the %s." and the vehicle part is not installed.

Comment (mugling) [inline src/veh_interact.cpp]: Noted. I've patched it to fix the instance @Coolthulhu noticed testing #17302 but I agree we should move the check higher to prevent future relapse.
